src/number_theory/legendre_symbol/mul_character.lean

```diff
@@ -0,0 +1,217 @@
+/-
+Copyright (c) 2022 Michael Stoll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Stoll
+-/
+import ring_theory.integral_domain
+
+/-!
+# Multiplicative characters of finite rings and fields
+
+Let `R` be a finite commutative ring.
+A *multiplicative character* of `R` with values in `R'` is a morphism of
+monoids with zero from the multiplicative monoid of `R` into that of `R'`.
+
+We use the namespace `mul_char` for the definitions and results.
+
+## Tags
+
+multiplicative character
+-/
+
+section multiplicative
+
+namespace mul_char
+
+/-!
+### Definitions related to and results on multiplicative characters
+-/
+
+universes u v w
+
+-- The domain of our multiplicative characters
+variables {R : Type u} [comm_ring R]
+-- The target
+variables {R' : Type v} [comm_ring R']
+
+/-- The trivial multiplicative character. It takes the value `0` on non-units and
+the value `1` on units. -/
+@[simps]
+def trivial [nontrivial R] [decidable_pred (λ x : R, is_unit x)] : R →*₀ R' :=
+{ to_fun := λ x, if is_unit x then 1 else 0,
+  map_zero' := by simp only [not_is_unit_zero, if_false],
+  map_one' := by simp only [is_unit_one, if_true],
+  map_mul' := by { intros x y,
+                   simp only [is_unit.mul_iff, boole_mul],
+                   split_ifs; tauto, } }
+attribute [protected] trivial
+
+/-- A multiplicative character takes unit values on units -/
+lemma is_unit_of_is_unit (χ : R →*₀ R') {a : R} (ha : is_unit a) : is_unit (χ a) :=
+begin
+  obtain ⟨u, rfl⟩ := ha,
+  exact is_unit_of_mul_eq_one _ _
+          (by rw [← map_mul, units.mul_inv, map_one] : χ u * χ (u⁻¹ : Rˣ) = 1),
+end
+
+/-- A multiplicative character is *nontrivial* if it takes a value `≠ 1` on a unit -/
+def is_nontrivial (χ : R →*₀ R') : Prop := ∃ (a : R), is_unit a ∧ χ a ≠ 1
+
+/-- A multiplicative character is *quadratic* if it takes only the values `0`, `1`, `-1` -/
+def is_quadratic (χ : R →*₀ R') : Prop := ∀ a, χ a = 0 ∨ χ a = 1 ∨ χ a = -1
+
+/-- Composition with an injective ring homomorphism preserves nontriviality -/
+lemma comp_is_nontrivial {χ : R →*₀ R'} (hχ : is_nontrivial χ) {R'' : Type w} [comm_ring R'']
+  (f : R' →+* R'') (hf : function.injective f) :
+   is_nontrivial (f.to_monoid_with_zero_hom.comp χ) :=
+begin
+  obtain ⟨a, ha₁, ha₂⟩ := hχ,
+  refine ⟨a, ha₁, λ ha, ha₂ (hf _)⟩,
+  erw [ha, map_one],
+end
+
+/-- Composition with a ring homomorphism preserves the property of being a quadratic character -/
+lemma comp_is_quadratic {χ : R →*₀ R'} (hχ : is_quadratic χ) {R'' : Type w} [comm_ring R'']
+  (f : R' →+* R'') : is_quadratic (f.to_monoid_with_zero_hom.comp χ) :=
+begin
+  intro a,
+  have hχ' : f.to_monoid_with_zero_hom.comp χ a = f (χ a) := rfl,
```

Comment:
I seemingly need it to make the rewrite possible. One can replace the lines `have hχ' : ...` and `rw [hχ']` by a `simp only` appealing to three simp lemmas, if this is preferable.

Reply:
I mean that your `hχ'` is [monoid_hom.comp_apply](https://leanprover-community.github.io/mathlib_docs/algebra/hom/group.html#monoid_hom.comp_apply) or something like that.

Reply:
Introducing a sublemma like `hχ'` is considered bad style for the following reason: imagine that one day someone will change the definition of composition, breaking the proof of your sublemma (this is of course very unlikely, but still). Then this person will have to fix it, and also fix all the place where something similar it's done. If instead everyone just uses `[monoid_hom.comp_apply](https://leanprover-community.github.io/mathlib_docs/algebra/hom/group.html#monoid_hom.comp_apply)`, then it will be enough to fix the proof of that lemma.

Reply:
(I have done this now).